File: cpp/binarySearch/koko.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

bool testar(vector<int>& piles, int k, int h) {
    double total = 0;
    if (k == 0) return false;
    for (int i = 0; i < piles.size(); i++) {
        total += (int)ceil((double)piles[i] / k);
        if (total > h) return false;
    }
    return true;
}
int minEatingSpeed(vector<int>& piles, int h) {

    int left = 1;
    int right = *max_element(piles.begin(), piles.end());
    while (left <= right) {
        int mid = (right + left) / 2;
        if (testar(piles, mid, h))
            right = mid - 1;
        else
            left = mid + 1;
    }
    return left;
}
int main() {

    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}

File: python/binarySearch/koko.py
import sys
import threading
import math
from typing import List, Tuple, Dict, Set
from collections import defaultdict, deque


class Solution:
    def testar(self, k, h, piles):
        if k == 0:
            return False
        total = 0
        for num in piles:
            total += math.ceil(num / k)
            if total > h:
                return False
        return True

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        res = max(piles)
        right = max(piles)
        while left <= right:
            mid = (right + left) // 2
            if self.testar(mid, h, piles):
                res = mid
                right = mid - 1
            else:
                left = mid + 1

        return res


def main():
    s = Solution()
    data = [1, 4, 3, 2]
    h = 9
    print(s.minEatingSpeed(data, h))


if __name__ == "__main__":
    threading.Thread(target=main).start()
